Remove ReduceOp::accumulator

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52196 Remove ReduceOp::accumulator**
* #52187 [nnc] Remove output_args from ReduceOp
* #52177 Avoid using ReduceOp->output_args() in rfactor
* #52170 Dont look at reduction output args when computing mem dependencies

Summary: A reduction does not need to know the buffer into which its
result will be written.  This change gets us closer to being able to
create reductions inside Compute, where we have access to the tensor
axes.

Test Plan: test_tensorexpr

Differential Revision: [D26420107](https://our.internmc.facebook.com/intern/diff/D26420107)

bertmaher committed:
Remove ReduceOp::accumulator

Summary: A reduction does not need to know the buffer into which its
result will be written.  This change gets us closer to being able to
create reductions inside Compute, where we have access to the tensor
axes.

Test Plan: test_tensorexpr

[ghstack-poisoned]

bertmaher committed:
clang-format on "Remove ReduceOp::accumulator"


Summary: A reduction does not need to know the buffer into which its
result will be written.  This change gets us closer to being able to
create reductions inside Compute, where we have access to the tensor
axes.

Test Plan: test_tensorexpr

Differential Revision: [D26420107](https://our.internmc.facebook.com/intern/diff/D26420107)

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
I like the direction these changes are going - I think they really make the representation of reductions cleaner! Let's just make sure that the potential issue with not-top-level reductions is documented (or even better - checked and guarded against) somewhere.

bertmaher commented:
> I like the direction these changes are going - I think they really make the representation of reductions cleaner! Let's just make sure that the potential issue with not-top-level reductions is documented (or even better - checked and guarded against) somewhere.

Yup, when I get to the point that will enable non-top-level reductions I'll add the appropriate guarding!

bertmaher committed:
fix clang-tidy on "Remove ReduceOp::accumulator"


Summary: A reduction does not need to know the buffer into which its
result will be written.  This change gets us closer to being able to
create reductions inside Compute, where we have access to the tensor
axes.

Test Plan: test_tensorexpr

Differential Revision: [D26420107](https://our.internmc.facebook.com/intern/diff/D26420107)

[ghstack-poisoned]
